[CORE-264] Server-rendered Job Search Check In Form

mellogarrett opened the pull request:
There's a lot going on here. I tried to include as much detail in the commit messages as possible. Locally this is pre-populating "Your Name" options correctly on the server.

I still need to write/fix some tests but wanted to get this out here for you all to have a look.

- Custom server to ensure the `Instruction` context has `.connect()`ed before routes start taking requests
- Handles pre-populating from both PostgREST and Forms API (snakecase and camelcase option keys & values)
- `prepopulateQuestions` has been updated so that it is "isomorphic" or can be ran on the server or the client. This means I didn't make `atlas-db` calls directly here, I went ahead and used `/GET` endpoint Josh & Mark created and am calling with an Axios client. There were a couple things I avoided by doing this that would've added complexity:
  - no need to worry about mocking `atlas-db` packages in the client with Webpack
  - the `req/res` objects available in `.getInitalProps` are different than the `req/res` objects available in `/api` routes meaning we'd need two different versions of the `getUserFromJwt` and `redirectToLogin` functions; one for client-side and one for server-side
- Since we aren't using `atlasClient` in `prepopulate-questions` and are instead using a plain Axios client, I updated the `get-prepopulate-url.ts` function to be a map of exact URLs rather than relative URLs. This means we can have Forms API & PostgREST prepopulate URLs co-existing here.
- When server-rendering, we can't redirect `302` to the login page so instead I'm returning `401` and then rendering a new `<Unauthorized/>` page instead with a link to the login page.

boblauer commented on pages/api/job-search-check-in.ts:
```diff
@@ -0,0 +1,27 @@
+    const user = await getUserFromJwt(req);
+    if (!user) return redirectToLogin(req, res);
+
+    if (req.method === 'POST') return submitForm(user.id, req.body).then(() => res.status(200).end());
```
The status here should be `201`, which is `Created`.

boblauer commented on pages/api/job-search-check-in.ts:
```diff
@@ -0,0 +1,27 @@
+
+async function submitForm(userId: string, data: object) {
+  const postData = (data as unknown) as SubmitJobSearchCheckInFormParams;
+  await atlas.instruction.submitJobSearchCheckInForm(userId, postData);
```
There's no reason to `await` here, it's the same as `return atlas...`.

mellogarrett replied:
We decided we don't want to return the created record back to the client here. We just want to wait until it's created and then return nothing.

boblauer replied:
We're ignoring the returned promise's value on L11 above, meaning the return value here won't affect the response we send to the server.

That said, if we really don't want to return anything from this function, can we add a `return null` after this to make it explicit?

boblauer commented on pages/api/job-search-check-in.ts:
```diff
@@ -0,0 +1,27 @@
+}
+
+async function handleGetRequest(userId: string) {
+  return (await atlas.instruction.getJobSearchCheckInForm(userId)) as object;
```
No reason to `return await` here, `return atlas...` achieves the same thing without the indirection.

boblauer commented on types/forms-api.ts:
```diff
@@ -0,0 +1,14 @@
+}
+
+export function isFormsApiOption(x: any): x is FormsApiOption {
+  return Object.keys(x).some((key) => {
```
Having a `some` with a `? false : true` after it hurts my brain. I think this is equivalent to

```js
return Object.keys(x).all(key => ['questionId', 'optionKey', 'optionValue', 'optionDetail'].includes(key));
```

boblauer commented on http/prepopulate-questions/index.test.ts:
```diff
@@ -1,22 +1,27 @@
 import { mockLogger } from '../../__mocks__/logger.mock';
+import axios from 'axios';
+import * as getPrepopulateUrl from '../../utils/get-prepopulate-url';
+import * as getDomain from '@lambdaschool/package-utils';
```
I would probably name this `as packageUtils` rather than `as getDomain`, since it's the full object and not just the `getDomain` function.

boblauer commented on package.json:
```diff
@@ -31,13 +32,15 @@
     "react-dom": "16.9.0",
     "react-ga": "2.5.7",
     "react-jss": "8.6.1",
+    "ts-node": "8.4.1",
```
Dev dependency?

boblauer commented on pages/daily-standup.tsx:
```diff
@@ -4,5 +4,4 @@ import { DailyStandupForm } from '../forms';
 
 const DailyStandupFormPage = () => <FormComponent pageTitle="Standup Form | Lambda School" form={DailyStandupForm} />;
 
-/* eslint-disable-next-line import/no-default-export */
```
I'm curious, why this is no longer necessary?

boblauer commented on utils/get-forms-api-url.ts/index.ts:
```diff
@@ -0,0 +1,12 @@
+import { FormTitle } from '../../constants/form-title';
+
+export const getFormsApiUrl = (formTitle: FormTitle) =>
```
Is this file used anywhere?

boblauer commented:
Lots of linting errors. You may want to configured your IDE to scream at you when you have a lint error, to avoid having to clean up after the fact.

MarkDKoslow commented on server/index.ts:
```diff
@@ -0,0 +1,22 @@
+import next from 'next';
+import { atlas } from '../domain/atlas';
+
+const portArg = process.argv[process.argv.findIndex((arg) => arg === '-p') + 1];
```
Ah this is clever

mellogarrett commented:
Agreed Mark. The added conditional stuff kinda stinks but I think we’ll be able to rethink it a bit once we migrate off like you mentioned.

boblauer commented on http/submit-form/backend-adapters/forms-api/index.ts:
```diff
@@ -0,0 +1,38 @@
+    formSubmission.merge_id = mergeId;
+  }
+
+  return form.questions.reduce((submission, question) => {
```
Looks like if there's no `externalKey` then we do nothing here, in which case we can break that apart into two steps:

```js
return form.questions
  .filter(question => question.externalKey)
  .reduce((submission, question) => Object.assign(submission, {
    [question.externalKey]: getAnswerForQuestion(question)
  }), formSubmission);
```

mellogarrett replied:
Actually, it looks like we can just remove that line. A `Question` will always have an `externalKey`.

mellogarrett commented:
True, although that means we'd be iterating over the same array twice which seems unnecessary to me.

boblauer commented:
I can't imagine we ever have more than 30 questions for a given form. Even if we had 1,000 questions, iterating over that array twice instead of once would maybe add 1ms, _maybe_.

I would always opt for smaller functions that do one thing over micro optimizations.

mellogarrett commented:
Alright, I believe this in a state to merge now. Any final reviews would be appreciated. 

I've tested locally and verified that authenticating, prepopulating, and submitting works on both client & server-rendered pages. It would be nice to do a similar test when this is deployed on Heroku.

boblauer commented on pages/job-search-check-in-server-rendered.tsx:
```diff
@@ -0,0 +1,20 @@
+    <FormComponent pageTitle="Job Search Check In | Lambda School" form={form} noResolver={true} />
+  ) : null;
+
+JobSearchCheckInPage.getInitialProps = async (ctx: NextPageContext) =>
```
Can do we something like this to achieve a 401 redirect?

```js
JobSearchCheckInPage.getInitialProps = async (ctx: NextPageContext) => 
{
	const props = await serverSidePrepopulateQuestions(JobSearchCheckInForm, ctx);
	if (!props.unauthorized) return props;

	res.writeHead(302, {
		Location: props.redirectAfterLoginUrl
	});
	res.end();
});
```

mellogarrett replied:
Possibly, let me play with this.

mellogarrett replied:
Alright, I was able to get this working 🎉

mellogarrett commented:
Made a pretty substantial refactor to the `http/prepopulate-questions/` folder (and also renamed it to `merge-question-options`) to get rid of all the conditional logic. This is now split out by backend.

Also, the redirect when unauthorized is working now so I removed the `<UnauthorizedComponent/>`

mellogarrett commented:
Agreed, sorry about that.
